huaweicloud/services/ddm/resource_huaweicloud_ddm_instance.go

```diff
@@ -0,0 +1,931 @@
+		"vpc_id":                utils.ValueIngoreEmpty(d.Get("vpc_id")),
+		"security_group_id":     utils.ValueIngoreEmpty(d.Get("security_group_id")),
+		"subnet_id":             utils.ValueIngoreEmpty(d.Get("subnet_id")),
+		"param_group_id":        utils.ValueIngoreEmpty(d.Get("param_group_id")),
+		"time_zone":             utils.ValueIngoreEmpty(d.Get("time_zone")),
+		"admin_user_name":       utils.ValueIngoreEmpty(d.Get("admin_user_name")),
+		"admin_user_password":   utils.ValueIngoreEmpty(d.Get("admin_user_password")),
+	}
+
+	return params
+}
+
+func buildCreateInstanceExtendParamChildBody(d *schema.ResourceData) map[string]interface{} {
+	params := map[string]interface{}{
+		"charge_mode":   utils.ValueIngoreEmpty(d.Get("charging_mode")),
+		"period_type":   utils.ValueIngoreEmpty(d.Get("period_unit")),
+		"period_num":    utils.ValueIngoreEmpty(d.Get("period")),
+		"is_auto_renew": utils.ValueIngoreEmpty(d.Get("auto_renew")),
+		"is_auto_pay":   "true",
+	}
+
+	return params
+}
+
+func resourceDdmInstanceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+
+	if d.HasChanges([]string{"name"}...) {
+		err := updateInstanceName(ctx, d, cfg, region)
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChanges([]string{"security_group_id"}...) {
+		err := updateInstanceSecurityGroup(ctx, d, cfg, region)
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChanges([]string{"flavor_id"}...) {
+		err := updateInstanceFlavor(ctx, d, cfg, region)
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChanges([]string{"node_num"}...) {
+		err := updateInstanceNodeNum(ctx, d, cfg, region)
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChanges([]string{"admin_user_password"}...) {
+		err := updateInstanceAdminPassword(ctx, d, cfg, region)
+		if err != nil {
+			return err
+		}
+	}
+	return resourceDdmInstanceRead(ctx, d, meta)
+}
+
+func updateInstanceName(ctx context.Context, d *schema.ResourceData, cfg *config.Config, region string) diag.Diagnostics {
+	// updateInstanceName: update DDM instance name
+	var (
+		updateInstanceNameHttpUrl = "v1/{project_id}/instances/{instance_id}/modify-name"
+		updateInstanceProduct     = "ddm"
+	)
+	updateInstanceNameClient, err := cfg.NewServiceClient(updateInstanceProduct, region)
+	if err != nil {
+		return diag.Errorf("error creating DDM client: %s", err)
+	}
+
+	updateInstanceNamePath := updateInstanceNameClient.Endpoint + updateInstanceNameHttpUrl
+	updateInstanceNamePath = strings.ReplaceAll(updateInstanceNamePath, "{project_id}", updateInstanceNameClient.ProjectID)
+	updateInstanceNamePath = strings.ReplaceAll(updateInstanceNamePath, "{instance_id}", fmt.Sprintf("%v", d.Id()))
+
+	updateInstanceNameOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		OkCodes: []int{
+			200,
+		},
+	}
+	updateInstanceNameOpt.JSONBody = utils.RemoveNil(buildUpdateInstanceNameBodyParams(d))
```

Comment:
I think here do not need `utils.RemoveNil()` function.

Reply:
done